src/server/req_gss_auth.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 1994-2019 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+/**
+ * @file	req_gsshandshake.c
+ *
+ * @brief
+ *  Batch request routines for GSS handshake via batch request (TCP).
+ */
+
+#include <pbs_config.h>   /* the master config generated by configure */
+
+#if defined(PBS_SECURITY) && (PBS_SECURITY == KRB5)
+
+#include "pbs_gss.h"
+#include "pbs_ifl.h"
+#include "log.h"
+#include "net_connect.h"
+#include "server.h"
+#include "dis.h"
+
+/* @brief
+ *	This should be called on a socket after obtaining client_name via
+ *	gss_accept_sec_context. It copies the credid from the client_name
+ *	structure to conn_credent.
+ *
+ * @param[in] s - tcp socket
+ *
+ * @return	int
+ * @retval	0 on success
+ * @retval	!= 0 on error
+ */
+int
+gss_set_conn(int s)
+{
+	char *credid;
+	int i, length;
+        conn_t *conn;
+	gss_extra_t *gss_extra = NULL;
+
+	/* this is done only once - after the gss handshake */
+	if (((gss_extra = (gss_extra_t *)tcp_get_extra(s)) != NULL) &&
+		gss_extra->establishing) {
+
+		conn = get_conn(s);
+
+		free(conn->cn_credid);
+		conn->cn_credid = strdup(gss_extra->clientname);
```

Comment:
where do we free() the allocated memory?

Reply:
In net_server.c:cleanup_conn()